std/mem.zig

```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
+    }
+
+    if (side == Side.RIGHT or side == Side.BOTH) {
+        while (end > begin and indexOfScalar(T, values_to_strip, slice[end - 1]) != null) : (end -= 1) {}
+    }
+
     return slice[begin..end];
 }
 
 test "mem.trim" {
-    assert(eql(u8, trim(u8, " foo\n ", " \n"), "foo"));
-    assert(eql(u8, trim(u8, "foo", " \n"), "foo"));
+    assert(eql(u8, trim(u8, " foo\n ", " \n", Side.BOTH), "foo"));
+    assert(eql(u8, trim(u8, "foo", " \n", Side.BOTH), "foo"));
+    assert(eql(u8, trim(u8, " foo ", " ", Side.LEFT), "foo "));
 }
 
 /// Linear search for the index of a scalar value inside a slice.
 pub fn indexOfScalar(comptime T: type, slice: []const T, value: T) ?usize {
-    return indexOfScalarPos(T, slice, 0, value);
+    return indexOfScalarPos(T, slice, 0, value, false);
 }
 
-pub fn indexOfScalarPos(comptime T: type, slice: []const T, start_index: usize, value: T) ?usize {
+pub fn indexOfScalarPos(comptime T: type, slice: []const T, start_index: usize, value: T, highest: bool) ?usize {
     var i: usize = start_index;
+    var out: ?usize = null;
+
     while (i < slice.len) : (i += 1) {
-        if (slice[i] == value)
-            return i;
+        if (slice[i] == value) {
+            out = i;
+            if (!highest) break;
+        }
     }
-    return null;
+
+    return out;
 }
 
 pub fn indexOfAny(comptime T: type, slice: []const T, values: []const T) ?usize {
-    return indexOfAnyPos(T, slice, 0, values);
+    return indexOfAnyPos(T, slice, 0, values, false);
 }
 
-pub fn indexOfAnyPos(comptime T: type, slice: []const T, start_index: usize, values: []const T) ?usize {
+pub fn indexOfAnyPos(comptime T: type, slice: []const T, start_index: usize, values: []const T, highest: bool) ?usize {
     var i: usize = start_index;
+    var out: ?usize = null;
+
     while (i < slice.len) : (i += 1) {
         for (values) |value| {
-            if (slice[i] == value)
-                return i;
+            if (slice[i] == value) {
+                out = i;
+                if (!highest) break;
+            }
         }
     }
-    return null;
+
+    return out;
 }
 
 pub fn indexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize {
-    return indexOfPos(T, haystack, 0, needle);
+    return indexOfPos(T, haystack, 0, needle, false);
 }
 
 // TODO boyer-moore algorithm
-pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, needle: []const T) ?usize {
+pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, needle: []const T, highest: bool) ?usize {
     if (needle.len > haystack.len)
         return null;
 
+    var out : ?usize = null;
     var i: usize = start_index;
     const end = haystack.len - needle.len;
     while (i <= end) : (i += 1) {
-        if (eql(T, haystack[i .. i + needle.len], needle))
-            return i;
+        if (eql(T, haystack[i .. i + needle.len], needle)) {
+            out = i;
+            if (!highest) break;
```

Comment:
see above comments

Reply:
Yep, I'll just reply once or twice as to not repeat myself.  I'll clean this up :).